src/api-utils.ts

```diff
@@ -51,6 +51,77 @@ export const backoffRetry = (attempts: number) => {
   return jitter(5000);
 };
 
+type RetryOptions<T> = {
+  timeout: number;
+  retryOn: (
+    attempts: number,
+    error: Error | null,
+    result: T | null,
+  ) => boolean | Promise<boolean>;
+  retryDelay: (attempts: number) => number;
+};
+
+/*
+ * helper function to perform an async operation where the caller can
+ * specify the retry and timeout semantics via an options API.
+ *
+ * in order for timeouts to be handled correctly, the contract with the caller is
+ * that the operation function must take an abort signal as an argument and
+ * must respect the signal by aborting the operation when the signal is aborted.
+ *
+ * in the case of a timeout, the `retryOn` function will be called with an error
+ * with the name "TimeoutError", which can be used to define how timeouts are retried.
+ */
+export const asyncOperationWithRetry = async <T>(
```

Comment:
Surprised we can't use something prebuilt for this functionality

Reply:
lmk if you know of a library you've used in the past, should be easy to drop it with the test coverage we have. when i looked everything i found either didn't quite have all of the features we'd want, or it was massive overkill (.e.g. `rxjs` or something like that)